GOcean OpenCL generates an in-kernel boundary mask to fix arbitrary problem sizes issues (closes #923)

sergisiso opened the pull request:
This PR adds a boundary mask as the first statements of the OpenCL kernels and passes the values of the boundaries as kernel arguments. (closes #923)

Additionally it adds a debug_mode that makes OpenCL easier to debug and increases the default local_size to 64 that performs much better in CPUs and GPUs

sergisiso commented:
Ready for the first review, probably from @arporter , it will also be good to know if this solves the 0 checksums in the PSyclone-generated NemoLite2D OpenCL that your workstation had.

arporter commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -433,15 +569,54 @@ def test_set_arg_const_scalar():
     ''' Tests that gen_ocl method of the GOcean Kernel Schedule generates
-    the expected OpenCL code.
+    the expected OpenCL code. Note this test doesn't prepare the kernel to
+    conform to OpenCL GOcean expected interface yet.
```
s/yet./: see test_opencl_prepared_kernel_code_generation for that./

sergisiso replied:
Done

arporter reviewed (changes requested):
Looks good. Just a few minor things to correct.
All tests pass including compilation.
I've tried this branch with master of PSycloneBench and the OCL version of (serial) generated nemolite2d still gives checksums of zero (on my laptop).

sergisiso commented on doc/user_guide/gocean1p0.rst:
```diff
@@ -855,6 +855,20 @@ small number of properties that must be defined with the right name:
+
+The GOcean configuration also includes a boolean parameter to enable or disable
+the generation of additional code which may impact performance but it is useful
+for debugging the application. By default is set to False, but it can be
```
Fixed

sergisiso commented on src/psyclone/configuration.py:
```diff
@@ -936,6 +938,15 @@ def __init__(self, config, section):
+                try:
+                    self._debug_mode = section.getboolean("debug_mode")
+                except ValueError as err:
+                    raise ConfigurationError(
```
Done

sergisiso commented on src/psyclone/configuration.py:
```diff
@@ -1029,6 +1040,17 @@ def grid_properties(self):
+    @property
+    def debug_mode(self):
+        '''
+        Getter for whether or not we generate additional debug code.
```
Removed line

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1110,22 +1106,43 @@ def gen_code(self, parent):
         '''
-        from psyclone.f2pygen import CallGen, DeclGen, AssignGen, CommentGen
-        # Create the array used to specify the iteration space of the kernel
+
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1110,22 +1106,43 @@ def gen_code(self, parent):
+        # Generate code to ensure data is on device
+        self.gen_data_on_ocl_device(parent)
+
+        # Extract the iteration boundary values which are given by the loops
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1110,22 +1106,43 @@ def gen_code(self, parent):
+        # surrounding the kernel, which won't be used in OpenCL. For now it
+        # converts the PSyIR to Fortran to use the values in f2pygen, but
+        # this should change in the future. Also note that OpenCL expects
+        # 0-indexing, hence we substract 1 to each boundary value.
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1178,14 +1208,50 @@ def gen_ocl(self, parent):
-                          "0", cnull, cnull])
+        if api_config.debug_mode:
+            # Check that everything is fine before the kernel launch
+            parent.add(AssignGen(parent, lhs=flag,
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1178,14 +1208,50 @@ def gen_ocl(self, parent):
+                parent, "check_status",
+                ["'Errors before {0} launch'".format(self.name), flag]))
+
+        args = ", ".join([
```
I was tired of having to search the OpenCL documentation every time I modified this line :)

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1286,6 +1379,86 @@ def gen_arg_setter_code(self, parent):
+
+    def _prepare_opencl_kernel_schedule(self):
+        ''' GOcean OpenCL kernels take the iteration boundaries as arguments
+        ans adds a conditional masking statement to avoid updating elements
```
Ops, fixed.

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1286,6 +1379,86 @@ def gen_arg_setter_code(self, parent):
+        '''
+        kschedule = self.get_kernel_schedule()
+
+        # If the first statement in the kernel schedule is not marked
```
Ops no, I dropped this functionality because even kernels that appear several times are just generated once in OpenCL, so the conditional mask check statement is always inserted just once as expected. I deleted the comment now.

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1286,6 +1379,86 @@ def gen_arg_setter_code(self, parent):
+        ystop_name = kernel_st.new_symbol_name("ystop")
+
+        # Create new symbols and insert them as kernel arguments after
+        # then initial iteration indices
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1139,17 +1156,30 @@ def gen_ocl(self, parent):
         # Then we set the kernel arguments
-        arguments = [kernel]
+        # In OpenCL the iteration boundaries are passed as arguments to the
+        # kernel because the group_work size may exceed the dimensions and
```
Fixed

sergisiso commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -88,24 +88,133 @@ def test_opencl_compiler_works(kernel_outputdir):
+        cmd_queues => get_cmd_queues()
+        kernel_compute_cu_code = get_kernel_by_name("compute_cu_code")
+      end if'''
+    print(generated_code)
```
Done

sergisiso commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -88,24 +88,133 @@ def test_opencl_compiler_works(kernel_outputdir):
+        CALL check_status("Global size is not a multiple of local size \
+(mandatory in OpenCL < 2.0).", -1)
+      END IF'''
+    print(generated_code)
```
Done

sergisiso commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -380,40 +502,54 @@ def test_set_kern_args_real_grid_property():
     argument. '''
     psy, _ = get_invoke("single_invoke_scalar_float_arg.f90", API, idx=0)
     sched = psy.invokes.invoke_list[0].schedule
+    sched.view()
```
Done

sergisiso commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -380,40 +502,54 @@ def test_set_kern_args_real_grid_property():
+    # This generated code cannot be compiled due to issue #798. Note the
+    # duplicated xstop symbol name in the argument list. This is not essential
+    # for the purpose of this test that just checks that the grid property
+    # xstop is declared as 'REAL(KIND=go_wp), intent(in), target :: a_scalar'
```
I reworded the comment. The problem is that there is an already existing xstop symbol which is not in the symbol table and therefore the OpenCL generation doesn't see the name clashing. Once that existing symbol is in the symbol table the "boundary xstop symbol" should automatically be "xstop_1".

sergisiso commented:
@arporter I addressed the comments, the PR is ready for the next review.

> I've tried this branch with master of PSycloneBench and the OCL version of (serial) generated nemolite2d still gives checksums of zero (on my laptop).

Can you try it once more but executing it with: `DL_ESM_ALIGNMENT=64 FORTCL_KERNELS_FILE=allkernels.cl ./nemolite2d.exe`
I see pocl has OpenCL1.2 and for OpenCL < 2  the global size being multiple of local size is mandatory.

arporter commented:
All requested changes have been made.
All tests pass/xfail (incl. compilation).
Examples all OK with compilation.
Reference and User guides build with no new warnings.
I'll bring up-to-date with master and let Travis report.

arporter commented:
This is ready to be merged but I'll do #955 first and then let @TeranIvy do #954.
